File: gccgo/py.go
// Copyright 2011 Julian Phillips.  All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package py

import (
	"fmt"
	"unsafe"
)

type _PyTypeObject __typeobject

func _malloc(int) unsafe.Pointer __asm__ ("malloc")
func _free(unsafe.Pointer) __asm__ ("free")

func __fopen(*int8,*int8) (*_FILE,int) __asm__ ("do_fopen")

func _fopen(name, mode *int8) (*_FILE, error) {
	f, errno := __fopen(name, mode)
	if f == nil {
		return nil, fmt.Errorf("Failed to open %s: %d", name, errno)
	}
	return f, nil
}

func _GoString(data *int8) string {
	s := uintptr(unsafe.Pointer(data))
	b := make([]byte, 0, 1024)
	for i := uintptr(0); *(*byte)(unsafe.Pointer(s + i)) != 0; i++ {
		b = append(b, *(*byte)(unsafe.Pointer(s + i)))
	}
	return string(b)
}

func _CString(s string) *int8 {
	b := _malloc(len(s)+1)
	data := []byte(s)
	for i, c := range data {
		x := (*byte)(unsafe.Pointer(uintptr(b) + uintptr(i)))
		*x = c
	}
	x := (*byte)(unsafe.Pointer(uintptr(b) + uintptr(len(s))))
	*x = 0
	return (*int8)(b)
}

File: gccgo/py.c
// Copyright 2011 Julian Phillips.  All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include "py.h"

#include "types.h"

extern void runtime_entersyscall(void) __asm__("syscall.Entersyscall");
extern void runtime_exitsyscall(void) __asm__("syscall.Exitsyscall");
extern void runtime_LockOSThread(void) __asm__("runtime.LockOSThread");
extern void runtime_UnlockOSThread(void) __asm__("runtime.UnlockOSThread");
extern int runtime_lockedOSThread(void);

extern int thread_locked(void) {
    return runtime_lockedOSThread();
}

void simple_cgocall(void (*f)(void*), void *a) {
    int lock = !runtime_lockedOSThread();
    if (lock) runtime_LockOSThread();
    runtime_entersyscall();
    f(a);
    runtime_exitsyscall();
    if (lock) runtime_UnlockOSThread();
}

void simple_cgocallback(void (*f)(void*), void *a) {
    runtime_exitsyscall();
    f(a);
    runtime_entersyscall();
}

void (*cgocallback)(void (*)(void*), void*) = simple_cgocallback;

#include "callbacks.c"
#include "utils.c"

extern PyObject *get__Py_NoneStruct(void) {
    return &_Py_NoneStruct;
}

extern PyTypeObject *get_PyBaseObject_Type(void) {
    return &PyBaseObject_Type;
}

extern PyObject **get_PyExc_Exception(void) {
    return &PyExc_Exception;
}

extern struct _dfr { FILE *f; int err; } do_fopen(const char *path, const char *mode) {
    struct _dfr s;
    s.err = 0;
    s.f = fopen(path, mode);
    if (!s.f) s.err = errno;
    return s;
}
